### PR TITLE
Add portal support to g_app_info_launch_default_for_uri

### PR DESCRIPTION
**Pull request by mariospr**
https://phabricator.endlessm.com/T12069

### CODE REVIEW

File: gio/gappinfo.c [hunk 1]
```diff
@@ -664,6 +668,68 @@ g_app_info_should_show (GAppInfo *appinfo)
   return (* iface->should_show) (appinfo);
 }
 
+static gboolean
+should_use_portal (void)
+{
+  const char *use_portal;
+  char *path;
+
+  path = g_strdup_printf ("/run/user/%d/flatpak-info", getuid());
```
Comment (cosimoc) [inline gio/gappinfo.c]: Use `g_get_user_runtime_dir()` instead?

File: gio/gappinfo.c [hunk 2]
```diff
@@ -664,6 +668,68 @@ g_app_info_should_show (GAppInfo *appinfo)
...
+launch_default_with_portal (const char         *uri,
+                            GAppLaunchContext  *context,
+                            GError            **error)
+{
+  GDBusConnection *bus;
+  GDBusMessage *message;
+  GVariantBuilder opt_builder;
+  const char *parent_window = NULL;
+
+  bus = g_bus_get_sync (G_BUS_TYPE_SESSION, NULL, NULL);
```
Comment (cosimoc) [inline gio/gappinfo.c]: This is leaked

File: gio/gappinfo.c [hunk 3]
```diff
@@ -664,6 +668,68 @@ g_app_info_should_show (GAppInfo *appinfo)
...
+    {
+      char **env;
+      env = g_app_launch_context_get_environment (context);
+      parent_window = g_environ_getenv (env, "PARENT_WINDOW_ID");
+      g_strfreev (env);
+    }
+
+  g_variant_builder_init (&opt_builder, G_VARIANT_TYPE_VARDICT);
+
+  message = g_dbus_message_new_method_call ("org.freedesktop.portal.Desktop",
```
Comment (cosimoc) [inline gio/gappinfo.c:726]: `message` is leaked

File: gio/gappinfo.c (hunk 2 above)
Comment (cosimoc) [inline gio/gappinfo.c]: No error checking?

File: gio/gappinfo.c [hunk 4]
```diff
@@ -664,6 +668,68 @@ g_app_info_should_show (GAppInfo *appinfo)
...
+  g_dbus_message_set_body (message, g_variant_new ("(ss@a{sv})",
+                                                   parent_window ? parent_window : "",
+                                                   uri,
+                                                   g_variant_builder_end (&opt_builder)));
+
+  g_dbus_connection_send_message (bus,
+                                  message,
+                                  G_DBUS_SEND_MESSAGE_FLAGS_NONE,
+                                  NULL,
+                                  NULL);
```
Comment (cosimoc) [inline gio/gappinfo.c]: Should do error checking here too

File: gio/gappinfo.c (hunk 1 above)
Comment (mariospr) [inline gio/gappinfo.c]: Makes sense

File: gio/gappinfo.c (hunk 2 above)
Comment (mariospr) [inline gio/gappinfo.c]: Right, I'll fix those in a follow up patch

File: gio/gappinfo.c (hunk 3 above)
Comment (mariospr) [inline gio/gappinfo.c:726]: Ditto

File: gio/gappinfo.c (hunk 4 above)
Comment (mariospr) [inline gio/gappinfo.c]: Ok

File: gio/gappinfo.c [hunk 5]
```diff
@@ -690,26 +723,27 @@ launch_default_with_portal (const char         *uri,
...
-  g_object_unref (session_bus);
+  g_dbus_message_set_body (message, g_variant_new ("(ss@a{sv})",
+                                                   parent_window ? parent_window : "",
+                                                   uri,
+                                                   g_variant_builder_end (&opt_builder)));
+
+  g_dbus_connection_send_message_with_reply (session_bus,
+                                             message,
+                                             G_DBUS_SEND_MESSAGE_FLAGS_NONE,
+                                             -1,
```
Comment (cosimoc) [inline gio/gappinfo.c]: Should be `G_MAXINT`

Comment (mariospr) [inline gio/gappinfo.c]: Ok, updated

**Pull request merged**